src/index.ts

```diff
@@ -281,14 +279,19 @@ export function register (options: Options = {}): Register {
   if (typeCheck) {
     // Set the file contents into cache.
     const setCache = function (code: string, fileName: string) {
-      cache.contents[fileName] = code
-      cache.versions[fileName] = (cache.versions[fileName] + 1) || 1
+      if (cache.contents[fileName] !== code) {
+        cache.contents[fileName] = code
+        cache.versions[fileName] = (cache.versions[fileName] + 1) || 1
+      }
     }
 
     // Create the compiler host for type checking.
     const serviceHost = {
       getScriptFileNames: () => Object.keys(cache.versions),
-      getScriptVersion: (fileName: string) => String(cache.versions[fileName]),
+      getScriptVersion: (fileName: string): any => {
```

Comment:
Also, can you make sure you fix the return type signature. It shouldn't be `any`.

Reply:
Unfortunately, here I can't - because TypeScript explicitly expects it to be [`(fileName: string): string`](https://github.com/Microsoft/TypeScript/blob/b94940525b77f31a8e7782c92fe3ea53371c93fb/src/services/types.ts#L165), if I make it `(fileName: string): string | undefined`, tsc will show an error of `getScriptVersion` type mismatch at the [line 315](https://github.com/TypeStrong/ts-node/blob/1d630f90f31ac7cc7f82b10145819e9405d65fd3/src/index.ts#L315).

As I said - the changes are controversial - I go against expected TypeScript interface, but that helps a lot with the performance because otherwise it will always think the codebase is outdated when comparing the files like `lib.d.ts`, which always have `undefined` version. It sucks I have to not conform the interface, but otherwise `ts-node` is unusable on my quite large project with type checking enabled...

Reply:
It starts as `undefined` (and stays `undefined`) for the files, which are not in [`config.fileNames`](https://github.com/TypeStrong/ts-node/blob/e7591c7e70c95152d8fd1380571ff870192c475e/src/index.ts#L226), but are in [`program. getSourceFiles()`](https://github.com/Microsoft/TypeScript/blob/592ee00906557ed313f7692fe6662251e65f915d/src/compiler/program.ts#L419).

As an experiment, if I change back `getScriptVersion` in `ts-node` to return a string always (`getScriptVersion: (fileName: string) => String(cache.versions[fileName])`), and then add some logging output to TypeScript, [there](https://github.com/Microsoft/TypeScript/blob/592ee00906557ed313f7692fe6662251e65f915d/src/compiler/program.ts#L419), to print out the files with non-matching versions:

```ts
// If any file is not up-to-date, then the whole program is not up-to-date
if (program.getSourceFiles().some(sourceFileNotUptoDate)) {
  console.log("");
  program.getSourceFiles().forEach(sourceFile => {
    if (sourceFile.version !== getSourceVersion(sourceFile.path)) {
      const printVersion = (str) => typeof str === "string" ? `"${str}"` : `${str}`;
      console.log(sourceFile.path, printVersion(sourceFile.version), printVersion(getSourceVersion(sourceFile.path)));
    }
  });
  return false;
}
``` 

I'll get something like this:

```
/users/anton/projects/my_project/node_modules/typescript/lib/lib.dom.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es2015.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es5.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es2015.symbol.wellknown.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es2015.reflect.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es2015.proxy.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es2015.iterable.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es2015.symbol.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es2015.promise.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es2015.generator.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es2015.collection.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/lib.es2015.core.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/chai/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react/global.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/redux/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/typescript/lib/typescript.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/walker/syntaxwalker.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/walker/walkcontext.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/walker/walker.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/walker/rulewalker.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/walker/scopeawarerulewalker.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/walker/blockscopeawarerulewalker.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/walker/programawarerulewalker.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/walker/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/rule/rule.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/configuration.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/formatter/formatter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/formatter/abstractformatter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatters/jsonformatter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatters/pmdformatter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatters/proseformatter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatters/verboseformatter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatters/stylishformatter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatters/fileslistformatter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatters/codeframeformatter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatters/tapformatter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatters/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatters.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/linter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/rule/abstractrule.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/rule/typedrule.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/rule/optionallytypedrule.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/rules.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/error.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/test/linterror.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/test.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/utils.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/ruleloader.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/enabledisablerules.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/formatterloader.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/language/utils.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/tslint/lib/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-dom/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/immutable/dist/immutable-nonambient.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/node/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/uuid/interfaces.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/uuid/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/lodash/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/mathjs/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/redux-thunk/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/reselect/src/reselect.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-dnd/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/redux-batched-actions/lib/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/prop-types/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/classnames/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-tabs/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-dnd-html5-backend/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/webfontloader/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/alpha/alpha.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/block/block.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/chrome/chrome.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/circle/circle.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/compact/compact.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/github/github.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/hue/hue.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/material/material.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/photoshop/photoshop.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/sketch/sketch.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/slider/slider.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/swatches/swatches.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/twitter/twitter.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/common/colorwrap.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/common/alpha.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/common/checkboard.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/common/editableinput.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/common/hue.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/common/saturation.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-color/lib/components/common/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/arrowkeystepper.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/autosizer.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/cellmeasurer.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/collection.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/columnsizer.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/list.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/table.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/grid.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/infiniteloader.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/masonry.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/multigrid.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/scrollsync.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/dist/es/windowscroller.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-virtualized/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/bowser/typings.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/exif-js/exif.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/xpath/xpath.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/xmldom/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/commander/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/parse5/lib/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/tough-cookie/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/jsdom/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/react-dom/server/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/cheerio/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/enzyme/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/sinon/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/redux-mock-store/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/arrify/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/chai-as-promised/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/chai-xml/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/chalk/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/diff/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/jquery/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/minimist/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/mkdirp/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/mocha/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/pako/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/sinon-chai/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/source-map-support/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/strip-bom/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/strip-json-comments/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/v8flags/index.d.ts "undefined" undefined
/users/anton/projects/my_project/node_modules/@types/yn/index.d.ts "undefined" undefined
```

I.e. all the `.d.ts` typings files from TypeScript stdlib and from node_modules always gonna have `undefined` version, and we'll compare it with `"undefined"` string.